Documentation for presigned URLs. (#90)
hms-dbmi/chromoscope: docs/presigned_urls.md

- list requirements.txt (i used a conda environment)

- what the overall function does: generates a presigned url for every object needed in the config file. makes the config file locally, uploads to config folder within S3, then generates a presigned URL for that new timestamped config file
- designed for: cohorts of samples with large amount of samples, with the same cancer type and assembly. good when comparing between samples within the same cohort (TODO: link out to dominika's stuff? will it be available? ask dom and sehi) 

- doesnt handle notes (ask if this is needed)
- samtools bai doesnt work on gunzipped file for me
- check imports
- terminal command to write to local file (> file.txt)
- print out to terminal in the script
- s3fs -- is this too much to mention?
- just exceptions? or raise other type of error?
- make public bucket for testing out?
- tell to check properties manually in cohort view
- generates samples and displays in order of ID list

    #TODO: create goscan URL and put in text file (including expiration date?) and upload to S3 (maybe use sync?)
    #TODO: look at gerstein github repo for example of walking through a script

>     # TODO: describe directory structure needed within the S3 bucket

Reviewer: This point is important. Please explain:
- what directory structure in the bucket is expected
- what file extensions are expected
- how to link files to sample IDs

Only [users with appropriate security credentials](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html#who-presigned-url) can create presigned URLs: [Identity and Access Management (IAM) users](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html), [IAM instance profiles](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html) (an IAM role for EC2 instances), and credentials generated via the AWS Security Token Service. The difference between these credentials presigned URLs generated with credentials for an (1) IAM instance profile is valid for up to 6 hours, (2) AWS Security Token Service is valid up to 36 hours when signed with permanent credentials, and (3) IAM user is valid up to 7 days. 

> For whichever set of credentials chosen, generate the appropriate AWS access ID-key pair. For the purposes of this documentation, generation of presigned URLs for an IAM user will be described, but follow AWS documentation to generate these keys for [EC2 instances](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html) and through the [AWS Security Token Service](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html). For IAM user profiles, a user-friendly way to generate access keys is using the [AWS console](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html#Using_CreateAccessKey), by navigating to the **Security credentials** tab under the appropriate user within the **IAM** (Identity and Access Management) area in the console and generating secret access keys. Save these values in the `~/.aws/credentials` file (`C:\Users\%USERPROFILE%\.aws\credentials` for Windows users). It will look similar to the following:

Reviewer 2: for the windows part. Has anyone from our team tried to set it up on Windows? If not, maybe it's too detailed

Author: I don't think anyone on our team has, since we're all using Macs to my knowledge. I'll remove this detail and let users of other OSes refer to the AWS docs

            "*"
        ],
        "AllowedMethods": [
            "GET"
        ],
        "AllowedOrigins": [
            "*"
        ],
        "ExposeHeaders": []
    }
]
```

&nbsp;
### AWS CLI Setup and Smoke Testing

Once the IAM credentials within an AWS account are set up, you can install the AWS CLI package and create presigned URLs. Follow the latest [AWS CLI documentation](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) to install the latest release of AWS CLI version 2 (or a past release of your choice).

Once installation is complete, you can check that the local AWS environment setup is functional using:

```
aws s3 ls <bucket_name>
```
which should list the files and subdirectories within the bucket `bucket_name` containing the data you would like to share via presigned URL. You can also use [*boto3*](https://aws.amazon.com/sdk-for-python/), the AWS SDK for Python, using the following script after [installing this SDK](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/quickstart.html):
```python
import boto3
bucket = s3.Bucket('<bucket_name>')
for bucket_obj in bucket.objects.all():
    print(bucket_obj.key)
```

> The latter will list *all* items within the given bucket recursively. *boto3* is also utilized within example scripts for creating presigned URLs within this repository, so it is recommended to install this SDK if employing these scripts are templates.

Reviewer 3: Seems to be a typo?

```suggestion
The latter will list *all* items within the given bucket recursively. *boto3* is also utilized within example scripts for creating presigned URLs within this repository, so it is recommended to install this SDK if employing these scripts as templates.
```

```
aws s3 presign s3://EXAMPLE_BUCKET/EXAMPLE_OBJECT --expires-in 604800
```
Example output:
```
https://EXAMPLE_BUCKET.s3.us-east-1.amazonaws.com/myobject?AWSAccessKeyId=AKIAEXAMPLEXXX&Expires=604800&Signature=ibOGfAovnhASUASsdasjj321dtg2s%3D
```
?> Note: When presigned URLs are created using the security credentials of an IAM user, the URL can last up to 7 days. When using a temporary token, the URL expires when the token expires, even if the specified time duration is longer.

For further optional arguments of this AWS CLI function, see the detailed documentation [here](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/presign.html).

One can also use AWS SDKs like *boto3* for Python to generate presigned URLs, using the `generate_presigned_url` function, using the [following script](../scripts/presigned_url_scripts/create_presigned_urls.py):
```python
import boto3
from botocore.exceptions import ClientError

> def generate_presigned_URL(bucket_name, object_path, expiration=3600):

Reviewer 2: I think, i's ok to show only the CLI approach since we want to present mainly  how to upload private data to the browser

    s3_client = boto3.client("s3")

    try:
        # URL generated by AWS user with access to private object (i.e. the owner)
        url = s3_client.generate_presigned_url(
            ClientMethod="get_object", # HTTP Method
            Params={
                "Bucket": bucket_name, # S3 bucket
                "Key": object_path # relative object path
            },
            ExpiresIn=expiration # URL duration in seconds
        )
    except ClientError as err:
        print(err)
        logging.error(err)
        return None

    # Return the presigned URL as a string
    return url
```
Further information on this *boto3* function can be found [here](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-presigned-urls.html).

Non-programmatically, a presigned URL can be generated using the S3 console or AWS Explorer for Visual Studio – [comprehensive documentation on presigned URLs can be found here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html). Once a presigned URL is created for a private object, it can be used within Chromoscope configuration files to temporarily visualize these private files.

&nbsp;

> ## Configuration File Creation Using Presigned URLs

Reviewer 3: Could be shortened to make it more readable on the sidebar

<img width="266" alt="Screenshot 2023-04-24 at 17 25 16" src="https://user-images.githubusercontent.com/9922882/234120287-9ba0b433-e112-4909-8e43-0ae39be14e3f.png">

### Presigned URLs within Chromoscope

> The ability to create presigned URLs not only allows for controlled, temporary visualization/sharing of private data, but also provides mode of sharing large amounts of data stored on the AWS cloud. Presigned URLs can be (1) used directly within a Chromoscope configuration file, linking to data for individual samples (simple single-sample example below): 

Reviewer 3: Would be good to link to our documentation

```suggestion
The ability to create presigned URLs not only allows for controlled, temporary visualization/sharing of private data, but also provides mode of sharing large amounts of data stored on the AWS cloud. Presigned URLs can be (1) used directly within a [Chromoscope configuration file](https://chromoscope.bio/docs/#/data-config?id=data-configuration), linking to data for individual samples (simple single-sample example below): 
```

```json
[
    {
        "id": "EXAMPLE_ID",
        "cancer": "breast",
        "assembly": "hg19",
        "sv": "https://EXAMPLE_BUCKET.s3.us-east-1.amazonaws.com/myobject.somatic.sv.bedpe?AWSAccessKeyId=AKIAEXAMPLEXXX&Expires=604800&Signature=alA%WOIDHCSha",
        "cnv": "https://EXAMPLE_BUCKET.s3.us-east-1.amazonaws.com/myobject.somatic.cna.annotated.tsv?AWSAccessKeyId=AKIAEXAMPLEXXX&Expires=604800&Signature=jj321dtg2s%3ibOGf"
    }
]
```

and/or (2) as a temporary URL for the configuration file itself (if it is saved as a private object within an S3 bucket), used as the argument for [the `external` parameter of a Chromoscope URL](url-parameters.md):
```
// format
https://chromoscope.bio/?external=[PRESIGNED_URL_FOR_YOUR_CONFIG_FILE]

// example
https://chromoscope.bio/?external=https://EXAMPLE_BUCKET.s3.us-east-1.amazonaws.com/myobject.configfile?AWSAccessKeyId=AKIAEXAMPLEXXX&Expires=604800&Signature=ibOGfAovnhASUASsdasjj321
```

&nbsp;

> ### Configuration File Creation Scripts: Overview

Reviewer 3: Maybe I missed it, but I wasn't sure if this generates a config file from already generated preassigned URLs of datasets or also generates preassigned URLs of datasets.

Author: generates presigned URLs for both the cohort data itself, and a presigned URL for the new configuration file it creates using those URLs

Author: ^ this is in the overview section of cohort config creation. let me know if this is still unclear

The latter script is designed to create Chromoscope configuration files for cohorts of samples, allowing for a user to compare samples within the same cohort visually using Chromoscope. There are several assumptions made for successful execution of this script, and will be expanded on below:
* Structure of the input TSV file containing sample IDs is as expected
* Directory structure in the S3 bucket containing the private data is as expected
* File formats for different kinds of data are of correct type, i.e. correct file extensions are used
* All samples are of the same **cancer type** and have the same **assembly**

&nbsp;
### Configuration File Creation Scripts: Sample ID List

> The first step completed when [generate_config_files.py](../scripts/presigned_url_scripts/generate_config_files.py) is executed is the definition of the list the sample IDs, from the TSV file passed into the parameter `ids`. These sample IDs leverage the expected structure of the subdirectories within the S3 bucket (see the following section) to link appropriate files to their corresponding sample within the configuration file, as well as define the `id` property for each sample within the configuration file.

Reviewer 3: You could add a link to the "following section" for easy navigtaion

The file used for this must be a tab-delimited file (TSV) that includes (but not limited to) a column named `ID`. The following example list, containing `n` samples:

```tsv
ID
SAMPLE_1_ID
SAMPLE_2_ID
...
SAMPLE_N_ID
```
will be converted to the following ID list:
```
[SAMPLE_1_ID, SAMPLE_2_ID,..., SAMPLE_N_ID]
```
Using this list, sample-wise configuration JSON objects are created iteratively, until all the objects are generated for all samples within the cohort. These IDs also correspond to subdirectories within the S3 bucket that contain that sample's files.

?> Note: Samples are added to the configuration file in the order they are listed in the ID TSV file. To change the ordering of the samples within Chromoscope, the order can be manually altered within the ID file. Refer to the [following documentation](../scripts/clustering/README.md) for an explanation of hierarchical clustering using the Euclidean metric as a distance measure.

&nbsp;

> ### Configuration File Creation Scripts: S3 Bucket Subirectory Structure

Reviewer 3: ```suggestion
### Configuration File Creation Scripts: S3 Bucket Subdirectory Structure
```

The configuration file creation script relies on an expected subdirectory structure within the S3 bucket, which is then accessed, sample-wise, according to the list of sample IDs from the submitted sample ID tsv file.

If there are several cohorts within the S3 bucket containing the private data, each should have their own subdirectory:
```bash
EXAMPLE_S3_BUCKET/
├── COHORT_1/
├── COHORT_2/
├── ...
└── COHORT_N/
```
Otherwise, if there is only one cohort within the bucket, ignore the above.

For the following level, the data should be subdivided into subdirectories corresponding to the following categories, with certain filetypes grouped together by sample (according the [data formats](data-formats.md) documentation):
| Subdirectory | Number of files (per sample) | Filetypes | Required |
|---|---|---|---|
| Structural variants (SV) | 1 | `bedpe` | Yes |
| Copy number variants (CNV) | 1 | `tsv` | Yes |
| Driver mutations | 1 | `tsv`, `json` | No |
| Point mutations (SNV) | 2 | `vcf` + `tbi` | No |
| Insertion/deletion mutations (Indel) | 2 | `vcf` + `tbi` | No |
| Read alignments | 2 | `bam` + `bai` | No |

> | Configuration files | - | `json` | No |

Reviewer 3: This is broken on my local machine.

<img width="1856" alt="Screenshot 2023-04-24 at 17 41 20" src="https://user-images.githubusercontent.com/9922882/234122934-2695e587-7222-4f2a-94eb-8a41ec8dbc7c.png">

└── CONFIGS_SUBDIR/                          # Contains timestamped configuration files (json)
    ├── example_config_a.json
    ├── example_config_b.json
    ├── ···
    └── example_config_z.json
```

?> The only categories that require subdirectories for *all* samples within the sample ID list are those for `SNV`s and `CNV`s. All other categories may contain some, all, or no sample subdirectories. If there is no sample data for a non-required category, that subdirectory need not be created.

?> The `configuration files` subdirectory, if not provided, is by default identified by the name **"CONFIGS"**. If it does not exist within the bucket, a new directory with this name will be created, and generated configuration files added to it. The configuration files are named by the time of their creation.

?> Note: If pairwise files don't pair correctly due to user error (e.g. a pair not of the same sample, not of correct file format), that category/parameter for that sample will not render within Chromoscope. This can be checked by manually inspecting samples within [cohort view](how-to-use.md#cohort-view).

&nbsp;
### Configuration File Creation Scripts: Usage

> The usage of the configuration file creation script via command line is as follows:

Reviewer 2: We should provide information about the used packages and their versions: `pandas` and `boto3`.

```

> python3 generate_config_files.py --ids IDS --bucket BUCKET [--cohort COHORT] --cancer CANCER --assembly ASSEMBLY --sv SV --cnv CNV [--drivers DRIVERS] [--snv SNV] [--indel INDEL] [--reads READS] [--configs CONFIGS] [--expiration EXPIR]

Reviewer 2: Maybe I missed it, but if not.. we should point the reader where the script lives in the repo

| `cnv` | - | Required. The name of the subdirectory within the S3 bucket containing copy number variant (CNV) `tsv` files. | `cnv` |
| `drivers` | `None` | Optional. The name of the subdirectory within the S3 bucket containing driver mutation `tsv` or `json` files. | `drivers` |
| `snv` | `None` | Optional. The name of the subdirectory within the S3 bucket containing point mutation (SNV) `vcf` files and their corresponding index `tbi` files. | `vcf`, `vcfIndex` |
| `indel` | `None` | Optional. The name of the subdirectory within the S3 bucket containing indel `vcf` files and their corresponding index `tbi` files.  | `vcf2`, `vcf2Index` |
| `reads` | `None` | Optional. The name of the subdirectory within the S3 bucket containing read alignment `bam` files and their corresponding index `bai` files. | `bam`, `bamIndex` |
| `configs` | `'CONFIGS'` | Optional. The name of the subdirectory, both locally and within the S3 bucket, that will contain the generated Chromoscope configuration `json` file. | - |
| `expiration` | `3600` | Optional. The duration of the presigned URLs within and for the generated Chromoscope configuration file, in seconds. | - |

> ?> The `note` property of the configuration file is not handled here, since it is a direct textual annotation. Only properties that require a URL value (in addition to required properties) are defined using this script.

Author: could load the note from the file. if it were a command line arg, it would be the same note for all samples, which is probably  not what most users would want -- for now I am still going to omit this and leave this note (for time's sake), and let developers modify the script themselves if they would like to

Only [users with appropriate security credentials](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html#who-presigned-url) can create presigned URLs: [Identity and Access Management (IAM) users](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html), [IAM instance profiles](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html) (an IAM role for EC2 instances), and credentials generated via the AWS Security Token Service. For whichever set of credentials chosen, generate the appropriate AWS access ID-key pair. For the sake of this documentation, generation of presigned URLs through an IAM user will be described, but follow AWS documentation to generate these keys for [EC2 instances](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html) and through the [AWS Security Token Service](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html).

> The difference between these credentials presigned URLs generated with credentials for an a) IAM instance profile is valid for up to 6 hours, b) AWS Security Token Service is valid up to 36 hours when signed with permanent credentials, and c) IAM user is valid up to 7 days. 

Reviewer 3: Just for the consistency with the paper.

```suggestion
The difference between these credentials presigned URLs generated with credentials for an (1) IAM instance profile is valid for up to 6 hours, (2) AWS Security Token Service is valid up to 36 hours when signed with permanent credentials, and (3) IAM user is valid up to 7 days. 
```

Note that presigned URLs grant access to S3 bucket data to whoever has the URL, so it is recommended to protect URL access and [limit URL capabilities](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html#PresignedUrlUploadObject-LimitCapabilities) appropriately. To allow visualization of data via presigned URLs for the SVELT browser, set the [cross-origin resource sharing (CORS) configuration file](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ManageCorsUsing.html) for the S3 bucket containing the private data to the following, [using the S3 console](https://docs.aws.amazon.com/AmazonS3/latest/userguide/enabling-cors-examples.html):

```
[
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "GET"
        ],
        "AllowedOrigins": [
            "*"
        ],
        "ExposeHeaders": []
    }
]
```

> jn

Reviewer 3: ```suggestion

```

-------

By default, all S3 objects are private. Only the object owner has permission to access them. However, the object owner can optionally share objects with others by creating a presigned URL, using their own security credentials, to grant time-limited permission to download the objects.

> When you create a presigned URL for your object, you must provide your security credentials and then specify a bucket name, an object key, an HTTP method (GET to download the object), and an expiration date and time. The presigned URLs are valid only for the specified duration. If you created a presigned URL using a temporary token, then the URL expires when the token expires, even if the URL was created with a later expiration time.

Author: great idea. I'll include in final draft.
